ftml/src/tree/element.rs

```diff
@@ -195,3 +196,126 @@ impl slog::Value for Element<'_> {
         serializer.emit_str(key, self.name())
     }
 }
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub enum Elements<'t> {
+    Multiple(Vec<Element<'t>>),
+    Single(Element<'t>),
+    None,
+}
+
+impl Elements<'_> {
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        match self {
+            Elements::Multiple(elements) => elements.is_empty(),
+            Elements::Single(_) => false,
+            Elements::None => true,
+        }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        match self {
+            Elements::Multiple(elements) => elements.len(),
+            Elements::Single(_) => 1,
+            Elements::None => 0,
```

Comment:
So `Elements` is a wrapper for either a list of elements, a single element (but not in a list) or nothing? Why this instead of  compositing a list that can be empty?

Reply:
You are correct that a `Vec<Element<'t>>` could serve the same purpose logically in the code.

However `Vec` allocates space on the heap (except if it's empty). Given that the most common result for a rule is returning a single element, it's a bit wasteful to always allocate when we could just pass the object on the stack. So with this enum, we only allocate if it's a rule that actually returns more than one element, and needs to have dynamic sizing.